src/internal_modules/roc_sndio/wav_source.cpp

```diff
@@ -0,0 +1,299 @@
+    return valid_;
+}
+
+bool WavSource::open(const char* path) {
+    roc_panic_if(!valid_);
+
+    roc_log(LogInfo, "wav source: opening: path=%s", path);
+
+    if (file_opened_) {
+        roc_panic("wav source: can't call open() more than once");
+    }
+
+    if (!setup_names_(path)) {
+        return false;
+    }
+
+    if (!open_()) {
+        return false;
+    }
+
+    if (!setup_buffer_()) {
+        return false;
+    }
+
+    file_opened_ = true;
+    return true;
+}
+
+DeviceType WavSource::type() const {
+    return DeviceType_Source;
+}
+
+DeviceState WavSource::state() const {
+    roc_panic_if(!valid_);
+
+    if (paused_) {
+        return DeviceState_Paused;
+    } else {
+        return DeviceState_Active;
+    }
+}
+
+void WavSource::pause() {
+    // no-op - but the state is updated
+    paused_ = true;
+}
+
+bool WavSource::resume() {
+    // no-op - but the state is updated
+    paused_ = false;
+    return true;
+}
+
+bool WavSource::restart() {
+    roc_panic_if(!valid_);
+
+    roc_log(LogDebug, "wav source: restarting: input=%s", input_name_.c_str());
+
+    if (!seek_(0)) {
+        roc_log(LogError, "wav source: seek failed when restarting: input=%s",
+                input_name_.c_str());
+        return false;
+    }
+
+    paused_ = false;
+    eof_ = false;
+
+    return true;
+}
+
+audio::SampleSpec WavSource::sample_spec() const {
+    roc_panic_if(!valid_);
+
+    if (!file_opened_) {
+        roc_panic("wav source: sample_spec(): non-open output file or device");
+    }
+
+    if (wav_.channels == 1) {
+        return audio::SampleSpec(size_t(wav_.sampleRate), audio::ChanLayout_Surround,
+                                 audio::ChanOrder_Smpte, audio::ChanMask_Surround_Mono);
+    }
+
+    if (wav_.channels == 2) {
+        return audio::SampleSpec(size_t(wav_.sampleRate), audio::ChanLayout_Surround,
+                                 audio::ChanOrder_Smpte, audio::ChanMask_Surround_Stereo);
+    }
+
+    roc_panic("wav source: unsupported channel count");
+}
+
+core::nanoseconds_t WavSource::latency() const {
+    return 0;
+}
+
+bool WavSource::has_latency() const {
+    roc_panic_if(!valid_);
+
+    if (!file_opened_) {
+        roc_panic("wav source: has_latency(): non-open input file or device");
+    }
+
+    return false;
+}
+
+bool WavSource::has_clock() const {
+    roc_panic_if(!valid_);
+
+    if (!file_opened_) {
+        roc_panic("wav source: has_clock(): non-open input file or device");
+    }
+
+    return false;
+}
+
+void WavSource::reclock(core::nanoseconds_t timestamp) {
+    // no-op
+    (void)timestamp;
+}
+
+bool WavSource::read(audio::Frame& frame) {
+    roc_panic_if(!valid_);
+
+    if (paused_ || eof_) {
+        return false;
+    }
+
+    if (!file_opened_) {
+        roc_panic("wav source: read: non-open input file");
+    }
+
+    audio::sample_t* frame_data = frame.samples();
+    size_t frame_left = frame.num_samples();
+
+    audio::sample_t* buffer_data = buffer_.data();
```

Comment:
Under the hood drwav uses fread() which has buffering, so no need for our own buffer on top.